### PR TITLE
add GRPC E2E support for more reducers

### PR DESCRIPTION
**Pull request by walterddr**
## Description
Support other GRPC reoducers

1. introduced streaming reducer and datatable reducer concept and use base reducers to share comment implementation codes.
2. added distrinct / groupby / aggregate reducers (gap-fill / selection order-by / explain plan yet to be implemented).

TODO:
adding both GRPC / Netty to broker (e.g. 2 different broker handlers runningon different ports) - need to change transport layer (`ServerRoutingInstance` and related classes) to make it support this multi-port paradigm. will be done in next PR.

### CODE REVIEW

File: pinot-core/src/main/java/org/apache/pinot/core/query/reduce/StreamingAggregationDataTableReducer.java [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+    _dataTableReducerContext = dataTableReducerContext;
+  }
+
+  /**
+   * Reduces data tables and sets aggregations results into
+   * 1. ResultTable if _responseFormatSql is true
+   * 2. AggregationResults by default
+   */
+  @Override
+  public void reduce(ServerRoutingInstance key, DataTable dataTable) {
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/query/reduce/StreamingAggregationDataTableReducer.java:77]: (major) We need to make sure all `reduce()` is thread safe. Same for other streaming reducers.

**Review by Jackie-Jiang — COMMENTED** [review]
Out of the scope of this PR, it might make sense to share the same reducer for both streaming and non-streaming query, but just use different reduce service. A lot of code is duplicated here

File: pinot-core/src/main/java/org/apache/pinot/core/query/reduce/StreamingAggregationDataTableReducer.java (hunk 1 above)
Comment (walterddr) [reply pinot-core/src/main/java/org/apache/pinot/core/query/reduce/StreamingAggregationDataTableReducer.java:77]: lol yeah this PR is not ready for review yet. just opening this draft to run CI :-P

File: pinot-core/src/main/java/org/apache/pinot/core/query/reduce/StreamingSelectionOnlyReducer.java [hunk 2]
```diff
@@ -35,8 +35,8 @@
 import org.slf4j.LoggerFactory;
 
 
-public class SelectionOnlyStreamingReducer implements StreamingReducer {
-  private static final Logger LOGGER = LoggerFactory.getLogger(SelectionOnlyStreamingReducer.class);
+public class StreamingSelectionOnlyReducer implements StreamingReducer {
```
[the remark this answers is not in the document]
Comment (walterddr) [reply pinot-core/src/main/java/org/apache/pinot/core/query/reduce/StreamingSelectionOnlyReducer.java]: done

File: pinot-core/src/main/java/org/apache/pinot/core/query/reduce/ResultReducerFactory.java [hunk 3]
```diff
@@ -68,11 +67,30 @@ public static DataTableReducer getResultReducer(QueryContext queryContext) {
   }
 
   public static StreamingReducer getStreamingReducer(QueryContext queryContext) {
-    if (!QueryContextUtils.isSelectionQuery(queryContext) || queryContext.getOrderByExpressions() != null) {
-      throw new UnsupportedOperationException("Only selection queries are supported");
+    AggregationFunction[] aggregationFunctions = queryContext.getAggregationFunctions();
+    if (aggregationFunctions == null) {
+      if (queryContext.getOrderByExpressions() != null) {
+        throw new UnsupportedOperationException("Only selection queries are supported");
```
[the remark this answers is not in the document]
Comment (walterddr) [reply pinot-core/src/main/java/org/apache/pinot/core/query/reduce/ResultReducerFactory.java:85]: there's select order by, gap-fill group by, and explain plan operator that hasn't been implemented.

**Comment by walterddr** [comment]
closing for now. @Jackie-Jiang and I discussed offline and felt like we don't really need the streaming processing in the non-selection-only situation. we will close this and create an issue to track the progress

**Pull request closed without merging**